Add Command Palette support

priethor opened the pull request:
##  What

This PR adds support for WordPress Command Palette/Core Commands (Cmd+K / Ctrl+K interface) by registering:
- Core commands to navigate to SCF admin pages
- "View All" commands for each Custom Post Type registered with SCF
- "Add new" for each Custom Post Type registered with SCF

## Implementation details

The commands are divided into admin and CPT command files.
- Core commands are only registered when the current user has admin permissions
- CPT commands are registered only when these conditions are met:
  - The current user has enough capabilities to edit the CPT
  - The CPT has enabled the `show_ui` setting

To prevent errors and optimize performance, JavaScript files for core and/or CPT commands will only be enqueued when needed.

##  Testing Instructions

  1. Ensure you're using WordPress 6.3+ (Commands API was introduced in 6.3)
  2. Ensure you have registered at least one CPT (e.g., "Books")
  3. Press Cmd+K (Mac) or Ctrl+K (Windows) in the Block Editor
  4. Type "SCF" or "Taxonomies" to see SCF-related admin commands
  5. Type your CPT name ("Books") to see its related commands ("View books", "Add book")
  6. Verify that commands correctly navigate to their respective pages
  7. As an administrator, verify you can see both admin commands and CPT ones
  8. As a user with limited permissions, verify that you only see commands you have access to. For example, an `editor` should only see CPT commands
  9. Test with a WordPress version earlier than 6.3 to confirm that no errors occur

## Recording

https://github.com/user-attachments/assets/fb85b69c-fd45-49e4-9918-2c39cc17f50c


## Followups
- Consider registering commands with `useCommandLoader` instead, which was introduced in WP 6.4.
- Add tests
- Implement a new REST endpoint for CPTs to avoid using `acf_localize_data`
- Migrate to TypeScript

priethor committed:
First attempt at adding command palette support

priethor committed:
Command polishing

priethor committed:
Add commands for registered CPTs

priethor committed:
Refactor

priethor committed:
Refactor to use wp.scf

priethor committed:
Refactor

priethor committed:
Refactor to pass the data in acf.data.customPostTypes

priethor committed:
Fix redirections

priethor committed:
Check capabilities before registering comands

Includes a refactor to divide commands between core and post-type based: core commands require admin permissions, whereas post-types are dynamic and depend on their own capabilities.

priethor committed:
Filter post types that don't enable the setting "show in UI"

priethor committed:
Redo webpack config linting

priethor committed:
Redo webpack linting

priethor committed:
Apply linting

priethor committed:
Polish

priethor committed:
Refactor

priethor committed:
Fix coding standards in comments

priethor committed:
Ensure we only register the commands when the palette is available

priethor committed:
Refactor to remove scf and acf prefixes, organizing files in folders similar to modules

priethor committed:
Optimize return early

gziolo commented on assets/src/js/commands/admin-commands.js:
```diff
@@ -0,0 +1,179 @@
+ * @since 6.5.0
+ */
+
+wp.domReady( () => {
```
It's possible to avoid using `domReady` and set the strategy for the registered script to `defer`. However, I see that `defer` [got introduced in WordPress 6.3](https://make.wordpress.org/core/2023/07/14/registering-scripts-with-async-and-defer-attributes-in-wordpress-6-3/), but the plugin supports WordPress 6.0 and higher.

priethor replied:
I can migrate it to use `defer`, since the JavaScript files won't be enqueued in versions earlier than 6.3 anyway, as 6.3 is when the Commands API was also introduced. 👍

gziolo replied:
> Ensure you're using WordPress 6.3+ (Commands API was introduced in 6.3)

So, technically speaking, this feature could be coded with `defer`.

> Consider registering commands with useCommandLoader instead, which was introduced in WP 6.4.

In that case, WP 6.3 users wouldn't be able to use this feature unless you provide 2 different implementations. What's the benefit of using `useCommandLoader`?

priethor replied:
> In that case, WP 6.3 users wouldn't be able to use this feature unless you provide 2 different implementations. What's the benefit of using useCommandLoader?

That's the main reason I didn't go directly with it, it's more straightforward to ensure the commands API is available. I think if we were to use `useCommandLoader`, we should only support that API and not support WP 6.3, having two implementations just for 6.3 seems too specific.

The Command Loader implementation is preferred because it's more performant, as instead of registering the command on page load, it defers the loading until the palette is used and only loads the specific commands that need to be shown to the user.

priethor replied:
In 887f2cf, I address this and optimize the scripts by:
- Deferring the load with the `defer` strategy.
- Deferring the execution by using the WP priority queue.

gziolo commented on assets/src/js/commands/admin-commands.js:
```diff
@@ -0,0 +1,179 @@
+		return;
+	}
+
+	const { __ } = wp.i18n;
+	const { createElement } = wp.element;
+	const { Icon } = wp.components;
+	const commandStore = wp.data.dispatch( 'core/commands' );
```
It is possible to use import statements to let the build tackle detecting all dependencies. @cbravobernal added all necessary code to automate it. See:
- https://github.com/WordPress/secure-custom-fields/pull/86

gziolo commented on assets/src/js/commands/admin-commands.js:
```diff
@@ -0,0 +1,179 @@
+			description: command.description,
+			keywords: command.keywords,
+			callback: ( { close } ) => {
+				document.location = command.url;
```
Nit: the following could help remove some duplication from the command config a bit:
```suggestion
				document.location = adminUrl + command.url;
```

gziolo commented on includes/assets.php:
```diff
@@ -243,6 +243,30 @@ public function register_scripts() {
 				);
 			}
 
+			// Register core scripts
```
There is most likely some code duplication as the same script handles are listed above. @cbravobernal should have more insights here. I also think that we could reuse the asset files generate in the build process together with `commands-admin` and `commands-custom-post-types`. I also belive this script handles need to be prefixed as it is only possible to register one script under a given script handle. These names are too general.

priethor replied:
Thanks for catching that, they are indeed a leftover after tests.

I've also prefixed the assets again with `scf-` to avoid potential collisions with other plugins.

gziolo commented on includes/assets.php:
```diff
@@ -262,6 +286,13 @@ public function register_scripts() {
 			 */
 			do_action( 'acf/register_scripts', $version, $suffix );
+
+			// Ensure WordPress scripts needed for command palette integration are available
+			if ( function_exists( 'wp_enqueue_script' ) ) {
+				wp_enqueue_script( 'wp-commands' );
+				wp_enqueue_script( 'wp-i18n' );
+				wp_enqueue_script( 'wp-dom-ready' );
+			}
```
This shouldn't be necessary if they are listed as dependencies for other enqueued scripts.

priethor replied:
Fixed!

gziolo reviewed (commented):
I left some initial general feedback before trying to better understand how it exactly works. I will try to test it next.

gziolo commented on assets/src/js/commands/admin-commands.js:
```diff
@@ -0,0 +1,179 @@
+
+	commands.forEach( ( command ) => {
+		commandStore.registerCommand( {
+			name: 'acf/' + command.name,
```
Nit: What do you think about the following?
```suggestion
			name: 'scf/' + command.name,
```

priethor replied:
Yes, that should be the right prefix. I'll fix it for the CPT commands as well.

gziolo commented:
You can also divide the work into two PRs since it’s fully functional and tests well. Command Palette support doesn’t have to be a Beta Feature because it’s a straightforward enhancement that anyone familiar with the concept should appreciate. 

Follow-up work could include:  
- Refactoring JS files to TS.
- Enhancing asset management to automatically detect dependencies in the build step.  

However, we need to take a closer look at the functionality, particularly the strategy for when these commands become available, depending on the WP version, the presence of custom post types, and user permissions.

priethor committed:
Remove unnecessary script registration

priethor committed:
Use the `scf` prefix in admin commands.

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

priethor committed:
Use `scf` prefix for CPT commands

gziolo commented on assets/src/js/commands/admin-commands.js:
```diff
@@ -0,0 +1,179 @@
+ */
+
+wp.domReady( () => {
+	if ( ! wp.data?.dispatch?.( 'core/commands' ) || ! acf?.data ) {
```
`@wordpress/commands` depends on `@wordpres/data` so if there is a check on the server whether `wp-commands` is registered then it should be enough as it ensure the rest works as intended. Do I miss some nuance here?

priethor replied:
Yep, this is specifically the `ACF` data. They are both defensive checks in case something didn't go well, but in theory, the commands API and `acf.data` should be available if the file is being executed.

gziolo commented on assets/src/js/commands/custom-post-type-commands.js:
```diff
@@ -0,0 +1,99 @@
+				'content',
+				'cpt',
+				postType.name,
+				postType.label || '',
```
Passing an empty string as a keyword should be avoided. Would it be possible to add it conditionally?

priethor replied:
Fixed!

gziolo commented on includes/admin/admin-commands.php:
```diff
@@ -0,0 +1,88 @@
+		}
+	}
+
+	acf_localize_data(
+		array(
+			'customPostTypes' => $custom_post_types,
+		)
+	);
+
+	if ( ! empty( $custom_post_types ) ) {
+		wp_enqueue_script( 'commands-custom-post-types' );
+	}
```
Nit, it feels like custom post types should be only passed when they are defined.
```suggestion
	if ( ! empty( $custom_post_types ) ) {
		acf_localize_data(
			array(
				'customPostTypes' => $custom_post_types,
			)
		);
		wp_enqueue_script( 'commands-custom-post-types' );
	}
```

priethor replied:
It's not a big difference (empty vs undefined), especially since the JS that uses this data will not be even enqueued, so let's do it as you suggest.

priethor committed:
Only pass CPTs to the frontend when the array is not empty

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

priethor committed:
CPT commands: only add label as a keyword if not empty

priethor committed:
Polish

priethor committed:
Refactor the depencies to user import statements

priethor committed:
Optimize scripts loading and execution

- Defer the script loading (defer was introduced in WP 6.3)
- Defer script execution by using the priority queue (introduced in WP 5.0)

Since the scripts will only be registered in versions with Command API support, which is 6.3+, we can use both

priethor committed:
Update package-lock.json

priethor committed:
Slight refactor

priethor committed:
Assets: remove redundant registrations, prefix registered scritps

youknowriad commented on assets/src/js/commands/admin-commands.js:
```diff
@@ -0,0 +1,193 @@
+/**
+ * Register admin commands for SCF
+ */
+queue.add( context, () => {
```
It's not really clear to me why you're using "priority-queue" here? Did you actually notice performance issues or something. Also potentially you can just use `requestIdleCallback` directly instead of using a queue with a single callback.

priethor replied:
I did not notice any performance difference; I implemented the priority queue in parallel to defer the loading [here](https://github.com/WordPress/secure-custom-fields/pull/121#discussion_r2060323077).

youknowriad replied:
Let's just get rid of it. I think `defer` is enough.

priethor replied:
I simplified to a `requestIdleCallback` here before reading your answer. Should I get rid of that, too? It still seems overkill.

youknowriad commented on assets/src/js/commands/admin-commands.js:
```diff
@@ -0,0 +1,193 @@
+	}
+
+	const commandStore = dispatch( 'core/commands' );
+	const adminUrl = window.acf?.data?.admin_url || '';
```
which script defines `window.acf` ?

cbravobernal replied:
The object is defined here:
https://github.com/WordPress/secure-custom-fields/blob/8b71adbd32aa0970c2c693a7c5e150ae2942289b/assets/src/js/_acf.js#L15

And the script is loaded here:
https://github.com/WordPress/secure-custom-fields/blob/8b71adbd32aa0970c2c693a7c5e150ae2942289b/includes/assets.php#L152

https://github.com/WordPress/secure-custom-fields/blob/8b71adbd32aa0970c2c693a7c5e150ae2942289b/includes/assets.php#L237

youknowriad commented on includes/admin/admin-commands.php:
```diff
@@ -0,0 +1,87 @@
+	}
+
+	// Ensure we only load our commands where the WordPress commands API is available.
+	if ( ! wp_script_is( 'wp-commands', 'registered' ) ) {
```
is this really necessary?

priethor replied:
This is necessary to avoid trying to access the command palette functions in WP versions earlier than 6.3 or without Gutenberg, in which cases my PR should do nothing or would result in errors.

youknowriad replied:
ok I didn't realize we support old versions like that in scf

youknowriad commented on includes/admin/admin-commands.php:
```diff
@@ -0,0 +1,87 @@
+ */
+function acf_commands_init() {
+	// Only load on admin screens.
+	if ( ! is_admin() ) {
```
This doesn't seem necessary since you're using `admin_queue_script`

priethor replied:
That's right, I removed it ✅

youknowriad commented on includes/admin/admin-commands.php:
```diff
@@ -0,0 +1,87 @@
+
+	$custom_post_types = array();
+
+	if ( function_exists( 'acf_get_acf_post_types' ) ) {
```
Why this check, can't we assume it exists since we're in the SCF codebase?

priethor replied:
There's an option to disable post type entirely, but checking the codebase confirms the function is loaded unconditionally (just returning empty if post types are disabled), so it is indeed safe to assume it always exists. 👍

youknowriad commented on includes/admin/admin-commands.php:
```diff
@@ -0,0 +1,87 @@
+	}
+
+	if ( ! empty( $custom_post_types ) ) {
+		acf_localize_data(
```
Personally, I'd argue that we should move away from localized variables like that as much as possible and instead use REST API endpoints. Do SCF has endpoints to retrieve these things, can't we just rely on Core endpoints for post types...?

cbravobernal replied:
I think the only endpoints registered are the ones that fetch custom post types and post fields, in every field you can chose to include it or not in the WP default API endpoints for posts fetching, but there are no internal endpoints to retrieve all post types or all post fields created with the plugin.

They are instead localizing the data: 

https://github.com/WordPress/secure-custom-fields/blob/8b71adbd32aa0970c2c693a7c5e150ae2942289b/includes/assets.php#L603

https://github.com/WordPress/secure-custom-fields/blob/8b71adbd32aa0970c2c693a7c5e150ae2942289b/includes/admin/post-types/admin-field-group.php#L151

Moving everything to a REST API could be done, but this needs to define a new API structure, with its own documentation and decide the approach for the store handling ( using Redux based like wp-data, or move to React Query or something similar )

What's your opinion on that @youknowriad ?

priethor replied:
There are two issues why using the Rest API would require some extra work and I didn't consider for this:
- While it does a PHP function to retrieve SCF Post types (`acf_get_acf_post_types`), it doesn't have a custom enpoint to do so, just WordPress` built in endpoints.
- More importantly, in order to show CPTs in the REST API, each CPT needs to enable the `Show in REST API` setting.

Therefore, my guess is we will need custom SCF endpoints for the admin that show all, as we will encounter a similar issue when implementing DataViews. What do you think about this approach, Riad?

youknowriad replied:
We're on WordPress, so for me, we should be use the entities and core-data here. no need to duplicate things that exist. 

> More importantly, in order to show CPTs in the REST API, each CPT needs to enable the Show in REST API setting.

For me this is something that we should try to get rid of. Every CPT should be visible in the endpoints, I think this is a setting of another age. 

That said, for now, a custom endpoint/entity could work if we're not ready to make all CPTs created by SCF show_in_rest true. Are there any valid reasons for not doing it?

priethor replied:
> Are there any valid reasons for not doing it?

If by this you mean a custom endpoint, the only reason not to do it right now is the extra complexity it adds to this PR, we should probably tackle that separately.

As per the existing setting, I wouldn't change an existing setting as we don't know all the cases or real users (e.g., having CPTs only for the admin to manage private data), though.

youknowriad replied:
> As per the existing setting, I wouldn't change an existing setting as we don't know all the cases or real users (e.g., having CPTs only for the admin to manage private data), though.

This is different than `show_in_rest`, this should be mostly handled by the `public` boolean.

priethor replied:
Thinking a bit more about this:
- Right now, we would need to use the `types` endpoint.
- `show_in_rest` determines whether a custom post type gets its own endpoint, but also hides the CPT definition from the `types` endpoint.
- In any case, the `types` endpoint doesn't return any information to know whether a type is a CPT and/or created by SCF (e.g., no `post-type` field)

Therefore, my inclination is to create a new endpoint for custom post types that ignores the `show_in_rest`. This will be needed, too, when migrating the Post Types screen to DataViews.

youknowriad replied:
The ideal scenario is to actually use the `types` endpoint and add "metadata" to the post type. That said, I think it's a good interim solution to have a custom endpoint.

youknowriad commented on includes/assets.php:
```diff
@@ -243,6 +243,28 @@ public function register_scripts() {
+			wp_register_script(
+				'scf-commands-admin',
+				acf_get_url( 'assets/build/js/commands/scf-admin' . $suffix . '.js' ),
+				array( 'acf', 'wp-plugins', 'wp-element', 'wp-components', 'wp-data', 'wp-commands', 'wp-i18n', 'wp-dom-ready' ),
```
why do these scripts depend on the "act" script?

cbravobernal replied:
I guess because they need the acf object to be there, as its being extended with the localized data.

priethor replied:
To maintain the proper script loading order and ensure things like `window.acf.data` already exist.

youknowriad replied:
I don't think we should be loading `acf` scripts in pages that might not need it. And the command palette itself can be rendered anywhere and ACF might not be loaded there. 

Do we need rely on `acf` just for the localization part? In other words, if we relied on an endpoint, we would need this right?

priethor replied:
Yep, if we relied on an endpoint we wouldn't need this. The only "real" dependency on ACF is to get the CPTs created by SCF from the DB.

youknowriad reviewed (commented):
In parallel to this PR, I wonder if we should implement [this PR](https://github.com/WordPress/gutenberg/pull/54515) within SCF (before it gets into Core) to enable the command palette everywhere.

priethor committed:
Remove redundant admin check

priethor committed:
Remove unnecesary checking of `acf_get_acf_post_types`

priethor committed:
Change the priority queue to a simpler `requestIdleCallback`

youknowriad commented:
So for me, the PR is fine. But I think we should ask ourselves what we want to do with these post types and any SCF UI going forward. Do we want to decouple the server and client and have a consistent data layer to communicate between both (ideally on top of core-data) or do we want to continue with these localized scripts which for me feels like hacks.

An investment in the core-data layer will help this PR but also any future PR that would like to rely on SCF entities in its client UI.

priethor commented:
> Do we want to decouple the server and client and have a consistent data layer to communicate between both (ideally on top of core-data) or do we want to continue with these localized scripts which for me feels like hacks.

I don't know if a hack, but it did feel like a dated approach that doesn't align with the current direction of core and modern architecture. I do agree with investing in decoupling server and client; the amount of data required in this PR is minimal, but as we move forward with new functionalities, it will pay off or even become necessary.

How do you all feel about this PR with the current approach, and then follow up with the new endpoint implementation?

youknowriad commented:
> How do you all feel about this PR with the current approach, and then follow up with the new endpoint implementation?

I'm ok with this personally, but I would try to avoid the `acf` script dependency, it seems this is not needed at all and we can just use any random global variable.

mcsf commented on includes/admin/admin-commands.php:
```diff
@@ -0,0 +1,80 @@
+		}
+
+		$plural_label   = $post_type['labels']['name'] ?? $post_type['label'] ?? $post_type['post_type'];
+		$singular_label = $post_type['labels']['singular_name'] ?? $post_type['singular_label'] ?? $plural_label;
```
Is `get_post_type_labels` not enough? If so, can we document why?

priethor replied:
Great catch, I've now leveraged that core function 👍

mcsf commented on includes/admin/admin-commands.php:
```diff
@@ -0,0 +1,80 @@
+		$plural_label   = $post_type['labels']['name'] ?? $post_type['label'] ?? $post_type['post_type'];
+		$singular_label = $post_type['labels']['singular_name'] ?? $post_type['singular_label'] ?? $plural_label;
+
+		$post_type_obj = get_post_type_object( $post_type['post_type'] );
```
I ask not knowing `acf_get_acf_post_types`: is it ever possible to have a key returned by `acf_get_acf_post_types` whose type object cannot be found? When?

priethor replied:
In theory it could in very edge cases, but we already checked `empty( $post_type['post_type'] )` [here](https://github.com/WordPress/secure-custom-fields/pull/121/files/71342ac399ed7169e8069cab83b248ea99a9335c#diff-4c2e11a2899b7dd644477d9386b324e9a81fb12be5db7bbee269a2a6338aa145R40), in which case this line wouldn't execute. The later checking of `$post_type_obj` also ensures this line executes properly.

mcsf commented on assets/src/js/commands/custom-post-type-commands.js:
```diff
@@ -0,0 +1,117 @@
+		}
+
+		const pluralLabel = postType.label || postType.name;
+		const singularLabel = postType.singular_label || pluralLabel;
```
Both the producer and the consumer of the `window.acf.data.customPostTypes` message are implementing their own fallback logic for computing these labels. Can we make it so that only one side (presumably, the producer) does that?

mcsf commented on assets/src/js/commands/custom-post-type-commands.js:
```diff
@@ -0,0 +1,117 @@
+			description:
+				/* translators: %s: Post type plural label */
+				sprintf(
+					__( 'SCF: View all %s', 'secure-custom-fields' ),
```
This is going to be a problem in i18n (stitching together fragments of a sentence from variables). Picture this sentence in Castilian and you'll see why: _Ver todos los_ (masc.) + _imágenes_ (fem.)

The answer in WordPress has generally been [expanding label support](https://developer.wordpress.org/reference/functions/get_post_type_labels/) for post types. The most adequate core label for your use case would be `view_items`, but I have no idea how widely it is implemented in SCF tradition, if at all, and its default value is _View Posts_, which will be confusing in this context.

You could argue that this is something for translators to solve for each locale, but I always worry when I see a string like `View all %s` that translators will either make mistakes or be left with an unreasonable burden to make it work in their locale.

How much worse would the UX be with these more i18n-friendly strings?
* `SCF: %s: View all entries` (more natural, IMO)
* `SCF: View all entries of type "%s"` (more robotic, IMO)
* same, but with `s/entries/items/`

Don't let this block you, btw, but give it a thought.

priethor replied:
I've tried a few approaches, and none seem ideal. I've proceeded with removing the prefixes and making the commands aligned with the wp-admin menu labels:
- Add command: It's now `Add %s`, which is also easier to translate.
- View all command: it's now `View all %s`. Still not ideal to translate, but it's the exact same string (and issue) as the admin menu label.

mcsf replied:
> Add command: It's now Add %s, which is also easier to translate.

Sure, in Castilian. ;) In more inflected languages, nouns may reflect their role in the sentence (e.g. nominative case for the subject, accusative for the direct object). In Esperanto:
- Page (nominative): _Paĝo_
- Add new page (accusative): _Aldoni novan paĝon_

> but it's the exact same string (and issue) as the admin menu label.

Not sure what you mean: are you saying it's the same string as the one shown in WP-Admin when you navigate the sidebar, e.g. _Posts -> All Posts_? If so, well then why not use the [actual label](https://core.trac.wordpress.org/browser/trunk/src/wp-includes/post.php?rev=60236#L2049)?

priethor replied:
Resolved [here](https://github.com/WordPress/secure-custom-fields/pull/121#issuecomment-2880698229)

mcsf commented on assets/src/js/commands/custom-post-type-commands.js:
```diff
@@ -0,0 +1,117 @@
+					__( 'SCF: View all %s', 'secure-custom-fields' ),
+					pluralLabel
+				),
+			keywords: [
+				'post type',
+				'content',
+				'cpt',
+				postType.name,
+				...( postType.label ? [ postType.label ] : [] ),
+			],
```
```suggestion
			keywords: [
				'post type',
				'content',
				'cpt',
				postType.name,
				postType.label,
			].filter( Boolean ),
```

mcsf commented on assets/src/js/commands/custom-post-type-commands.js:
```diff
@@ -0,0 +1,117 @@
+			label:
+				/* translators: %s: Post type singular label */
+				sprintf(
+					__( 'Add New %s', 'secure-custom-fields' ),
```
See my other comment about translating post labels. Here, it's `add_new` or `add_new_item`.

mcsf commented on assets/src/js/commands/custom-post-type-commands.js:
```diff
@@ -0,0 +1,117 @@
+			description:
+				/* translators: %s: Post type singular label */
+				sprintf(
+					__( 'SCF: Create a new %s', 'secure-custom-fields' ),
```
Same.

mcsf commented on assets/src/js/commands/custom-post-type-commands.js:
```diff
@@ -0,0 +1,117 @@
+				'create',
+				'content',
+				postType.name,
+				...( postType.label ? [ postType.label ] : [] ),
```
Same suggestion with `[].filter(Boolean)` on the keywords array

mcsf commented:
> So for me, the PR is fine. But I think we should ask ourselves what we want to do with these post types and any SCF UI going forward. Do we want to decouple the server and client and have a consistent data layer to communicate between both (ideally on top of core-data) or do we want to continue with these localized scripts which for me feels like hacks.

It's better in the long run to have "proper" communication, but — when we get to that — let's ensure that these payloads are preloaded, which is something we shouldn't take for granted in the old-school localize_script method.

priethor commented:
Thanks, I'll address the feedback and merge the PR; I just submitted [the required changes to the `types` endpoint](https://github.com/WordPress/secure-custom-fields/pull/128) to migrate the Command Palette from using `acf_localiza_data` to fetching the types from the REST API.

priethor committed:
Update assets/src/js/commands/custom-post-type-commands.js

Co-authored-by: Miguel Fonseca <150562+mcsf@users.noreply.github.com>

priethor committed:
Improved URL building

priethor committed:
Fix command labels for better i18n

priethor committed:
Simplify label calculation leveraging WP core functions

priethor committed:
Remove redundant client-side label calculation

priethor committed:
Merge remote-tracking branch 'origin/trunk' into add/command-palette-support

priethor committed:
Fixed `@since` annotations to follow the steps of #129

mcsf reviewed (approved):
We need to address the i18n issues right after this, but ship it 🚢

priethor committed:
Switch to using the proper labels for the commands

priethor commented:
> We need to address the i18n issues right after this, but ship it 🚢

Thanks for the feedback, based on our last chat, the last commit directly uses the `all_items` and `add_new_items` provided instead, so we don't add any translatable text and reuse the same labels as wp-admin. These already have fallbacks in case they are empty, too.
